### ISSUE TITLE
kola --parallel broken with qemu

### ISSUE THREAD

**Issue by cgwalters** [coreos/mantle#1157]
In my test FCOS pipeline, one of the nodes ran OOM.  Looking at it, we're spawning too many qemu instances.  If I do `cosa kola run -- --parallel 2` locally, I get 8 qemu proceses.

Yet, I only see 1 concurrent if I don't specify `--parallel`.  So the bug seems to occur when parallel > 1?

I started reading the harness code around parallelism but my eyes glazed over - so many goroutines and barriers and mutexes and lions and tigers and bears...

**Comment by cgwalters** [comment]
I don't think this is actually specific to qemu of course.  Probably we just don't notice it as much.

Hum...could this have something to do with subtests?

**Comment by cgwalters** [comment]
Ah I see, this code came from https://github.com/coreos/mantle/commit/d5f50b5ab0886b1953359e1b3ef816d49132fe94 - it forked from the golang upstream, so of course they're going to be all "channels solve everything!".

Ah, and https://github.com/golang/go/commit/f04d5836181dec3ec1b7e427607f02fa7a204a2d#diff-70e628298261565d825f7199d13042f2 looks relevant here.

And it's Go so of course there's a pile of race conditions with unstructured goroutines, like
- https://github.com/golang/go/commit/7f31971f594edbacbdba5407aaee042850fbd220#diff-70e628298261565d825f7199d13042f2
- https://github.com/golang/go/commit/eff1e68528fc9052a7ff1ac7afe222696f85db8c#diff-70e628298261565d825f7199d13042f2

Blah.

**Comment by arithx** [comment]
> If I do cosa kola run -- --parallel 2 locally, I get 8 qemu proceses.

Notably the `parallel` flag concerns test count not machine count so theoretically you could have a much larger machine count than the value of the parallel flag if the tests used require multiple machines (though in practice I don't believe we have a test with more than 3 machines, and it might be less now after dropping legacy CL tests).

> I started reading the harness code around parallelism but my eyes glazed over - so many goroutines and barriers and mutexes and lions and tigers and bears...

Yeah it's not pretty unfortunately, [the runner readme](https://github.com/coreos/mantle/blob/master/runner-readme.md) might help a bit.

**Comment by cgwalters** [comment]
So...hum.  How about `--qemu-parallel N` which ensures that at most N qemu Flight instances are created at a time?

**Comment by arithx** [comment]
> So...hum. How about --max-flights N which ensures that at most N Flight instances are created at a time?

Flights can have multiple clusters, and clusters can have multiple machines. If you're looking to keep to a particular maximum instance count it'll be very difficult as individual tests can spawn up additional machines outside of the declared amount (this is done in tests like [the NFS ones](https://github.com/coreos/mantle/blob/master/kola/tests/misc/nfs.go#L66) to allow them to create machines with different configs)

**Comment by jlebon** [comment]
> Yet, I only see 1 concurrent if I don't specify --parallel. So the bug seems to occur when parallel > 1?

Hmm, weird. I'm not even seeing that. I just did a simple `cosa kola` and ended up with 19 qemu processes! Looking closely it seems like the qemu processes aren't actually being reaped after their associated tests complete.

Also seeing this with e.g. `--parallel 4`.

**Comment by cgwalters** [comment]
> Looking closely it seems like the qemu processes aren't actually being reaped after their associated tests complete.

You see ones in `Z` state?  I'm not seeing that here.

**Comment by cgwalters** [comment]
> Flights can have multiple clusters, and clusters can have multiple machines. If you're looking to keep to a particular maximum instance count it'll be very difficult as individual tests can spawn up additional machines outside of the declared amount (this is done in tests like the NFS ones to allow them to create machines with different configs)

OK.  Well, we can add at least `--qemu-max-machines` right?

**Comment by cgwalters** [comment]
Or a generic `--max-machines` would be straightforward too I think, just a mechanical change.

**Comment by cgwalters** [comment]
PR in https://github.com/coreos/mantle/pull/1161

**Comment by cgwalters** [comment]
OK so yeah we were definitely leaking qemu instances.  And that may have been what was provoking a kernel bug at least on RHEL7 that was taking down the RHCOS pipeline.
